File: examples/cpp/read.cpp
#include <libxml/xmlreader.h>
#include <libxml/xinclude.h>

#include "xidx/xidx.h"

using namespace xidx;

#define IndexSpace3d IndexSpace<3>

int main(int argc, char** argv){

  if(argc < 2){
    fprintf(stderr, "Usage: read file_path [debug]\n");

    return 1;
  }

  clock_t start, finish;
  start = clock();

  XidxFile metadata(argv[1]);

  int ret = metadata.Load();

  finish = clock();

  printf("Time taken %fms\n",(double(finish)-double(start))/CLOCKS_PER_SEC);

  std::shared_ptr<Group> root_group = metadata.GetRootGroup();
  
  std::shared_ptr<Domain> time_domain = root_group->GetDomain();
  
  //DomainType dom_type = time_domain->type;
  
  std::shared_ptr<TemporalListDomain> domain = std::static_pointer_cast<TemporalListDomain>(time_domain);
  
  printf("Time Domain[%s]:\n", ToString(domain->type));
  
  for(auto& t : domain->GetLinearizedIndexSpace()){
    printf("Timestep %f\n", t);
    
    for(auto& grid: root_group->GetGroups()){
      std::shared_ptr<Domain> domain = grid->GetDomain();
      
      printf("\tGrid Domain[%s]:\n", ToString(domain->type));
      
      for(auto& var: grid->GetVariables()){
        printf("\t\tVariable: %s %s\n", var->name.c_str(), var->data_items[0]->GetXPath().c_str());
        
      }
    }
    
  }
  
  
//    auto it = domain->GetLogicalIterator();
//    
//    IndexSpace3d is = domain->GetIndexSpace<3>();
//    IndexSpaceIterator<3> isit(is);
//    Rect<3> r = isit.rect;
//    PointInRectIterator<3, PHY_TYPE> pit(r);
    
    
    
    
    
//
//    Rect<3> r = is.bounds;
    // for(auto idx: is)
    //   //IndexSpaceSelection iss(domain, idx);
    //   Group g = root_group->GetInstance(idx)
    //   IndexSpace gis = g.GetIndexSpace()
    //   vector<SubSpace>
  
  
  
  
  metadata.Save("verify.xidx");
  
  if (argc < 3) // if no args just load, otherwise debug
    return 0;

#if 0
  //typedef std::list<std::shared_ptr<XidxGroup> > XidxGroupList
  XidxGroup root = meta.getRootGroup();

  XidxDomain time_dom = root.getDomain();

  //print time_dom

  for(auto item : root.getGroups()){
    
    for(auto grid : item.getGroups()){
      XidxDomain dom = grid.getDomain();
      
      for(auto var : grid->getVariables()){

        // print var

        // get pointer to the data 
        var.getFilePointers();

        // printf("Variable %s cent %s type %s num_type %s prec %s endian %s DType %s Format %s\n",
        //   var.name.c_str(), ToString(att.centerType), ToString(att.attributeType),
        //   ToString(att.data.numberType), att.data.precision.c_str(), ToString(att.data.endianType), att.get_dtype_str().c_str(), ToString(att.data.formatType));

          // if(var.data.formatType == FormatType::BINARY_FORMAT){
          //   std::vector<int> idims;
          //   std::istringstream f(att.data.dimensions);

          //   std::string s;
          //   while (getline(f, s, ' ')) {
          //     idims.push_back(atoi(s.c_str()));
          //   }

          //   size_t size = 1;
          //   for(auto d:idims)
          //     size *= d;
          //   char data[size];
          //   att.get_raw_data(data,size);
          //   printf("\t<< reading binary attribute of size %zu\n", size);
          // }

        for(auto att: var.getAttributes()){
          printf("\t Info %s = %s \n", att.name.c_str(), att.value.c_str());
        }
      }

    }

  }

#endif

  // add a timestep to the current metadata
  // std::shared_ptr<TimeStep> ts(new TimeStep());
  // std::shared_ptr<Level> level = meta.get_timestep(0)->get_level(0);

  // ret = ts->add_level(level);

  // ts->set_timestep(777, float(7777));
  // meta.add_timestep(ts);

  // write the metadata to compare with original
  //meta.set_md_file_path("load/test_hpc.xmf");
  // meta.save();

  return ret;
}

File: examples/python/read.py
import xidx
import sys

from xidx import *

if(len(sys.argv) < 2):
  print("usage: ", sys.argv[0], "<input file>")
  exit(1)

filepath = sys.argv[1]

# create metadata file object
metadata = MetadataFile(filepath)
# read data from input file
ret = metadata.Load()

# get root group (e.g. Time series)
root_group = metadata.getRootGroup()

# get domain of the group
domain = root_group.getDomain()

print ("Time Domain[",Domain.toString(domain.type),"]:")

# print attributes if any
for att in domain.getAttributes():
  print ("\t\tAttribute:", att.name, "Value:", att.value)

# loop over the list of timesteps
for t in domain.getLinearizedIndexSpace():
  print ("Timestep", t)
    
  # loop over the groups (e.g. grids)
  for grid in root_group.getGroups():
    # get domain of current grid
    grid_domain = grid.getDomain()
    
    print ("\tGrid Domain[", Domain.toString(grid_domain.type), "]")

    # print attributes if any
    for att in grid_domain.getAttributes():
      print ("\t\tAttribute:", att.name, "Value:", att.value)

    if(grid_domain.type == Domain.SPATIAL_DOMAIN_TYPE):
      print ("\tTopology", Topology.toString(grid_domain.topology.type), "volume ", grid_domain.getVolume())
      print ("\tGeometry", Geometry.toString(grid_domain.geometry.type))
    elif(grid_domain.type == Domain.MULTIAXIS_DOMAIN_TYPE):
      # loop over the axis 
      for a in range(0,grid_domain.getNumberOfAxis()):
        # get axis
        axis = grid_domain.getAxis(a);
        print ("\tAxis", axis.name,"volume", axis.getVolume(),": [ ", end='')

        for v in axis.getValues():
          print (v, end='')

        print(" ]")

        # print attributes of the axis if any
        for att in axis.getAttributes():
          print ("\t\tAttribute:", att.name, "Value:", att.value)
  
      print("\n");
      
      # loop over variables
      for var in grid.getVariables():
        # get datasource used by the first variable
        source = var.getDataItems()[0].getDataSource();
        print("\t\tVariable:", var.name)
        if(source):
          print("\t\t\tdata source url: ", source.getUrl())
        else:
          print("\n")
        
        # print attributes of the variable if any
        for att in var.getAttributes():
          print ("\t\tAttribute:", att.name, "Value:", att.value)
        
# (Debug) Saving the content in a different file to compare with the original
metadata.save("verify.xidx");
print("debug saved into verify.xidx")


